text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+ - Repetition (zero or more times) is signified by curly braces (as in `{ <name-prefix> }`)
+ - Comments are marked with `//`.
+
+Mangled names conform to the following grammar:
+
+```
+// The <decimal-number> specifies the encoding version.
+<symbol-name> = "_R" [<decimal-number>] <absolute-path> [<instantiating-crate>]
+
+<absolute-path> = "N" <path-prefix> [<generic-arguments>] "E"
+                | <substitution>
+
+<path-prefix> = <identifier>
+              | "M" <type>
+              | "X" <type> <absolute-path> [<disambiguator>]
+              | <path-prefix> <identifier>
+              | <substitution>
+
+// The <decimal-number> is the length of the identifier in bytes.
+// <bytes> is the identifier itself and must not start with a decimal digit.
+// If the "u" is present then <bytes> is Punycode-encoded.
+// "V" and "C" are the tags for value and closure namespaces respectively.
+<identifier> = <decimal-number> <bytes> ["u"] ["V"|"C"] [<disambiguator>]
+
+<type> = <basic-type>
+       | <absolute-path>                  // named type
+       | "A" [<decimal-number>] <type>    // [T; N]
+       | "T" {<type>} "E"                 // (T1, T2, T3, ...)
+       | "R" <type>                       // &T
+       | "Q" <type>                       // &mut T
+       | "P" <type>                       // *const T
+       | "O" <type>                       // *mut T
+       | "G" <identifier> "E"             // generic parameter name
+       | <function-type>
+       | <substitution>
+
+<basic-type> = "a"      // i8
+             | "b"      // bool
+             | "c"      // char
+             | "d"      // f64
+             | "e"      // str
+             | "f"      // f32
+             | "h"      // u8
+             | "i"      // isize
+             | "j"      // usize
+             | "l"      // i32
+             | "m"      // u32
+             | "n"      // i128
+             | "o"      // u128
+             | "s"      // i16
+             | "t"      // u16
+             | "u"      // ()
+             | "v"      // ...
+             | "x"      // i64
+             | "y"      // u64
+             | "z"      // !
+
+// If the "U" is present then the function is `unsafe`.
+// If the "J" is present then it is followed by the return type of the function.
+<function-type> := "F" ["U"] [<abi>] {<type>} ["J" <type>] "E"
+
+<abi> = "K" (
+           "d" |   // Cdecl
+           "s" |   // Stdcall
+           "f" |   // Fastcall
+           "v" |   // Vectorcall
+           "t" |   // Thiscall
+           "a" |   // Aapcs
+           "w" |   // Win64
+           "x" |   // SysV64
+           "k" |   // PtxKernel
+           "m" |   // Msp430Interrupt
+           "i" |   // X86Interrupt
+           "g" |   // AmdGpuKernel
+           "c" |   // C
+           "x" |   // System
+           "r" |   // RustCall
+           "j" |   // RustInstrinsic
+           "p" |   // PlatformInstrinsic
+           "u"     // Unadjusted
+        )
```

Comment:
This all seems a bit arbitrary. Given that in principle there could be an unbounded number of ABIs, it seems like we should splurge on using a real string here rather than a single character. I'm also going to guess that these will be relatively rare, so space isn't a consideration?

Reply:
I'd also favor encoding the ABI "string" (ideally as an identifier, replacing `-` with `_`, etc.)

Reply:
I responded in https://github.com/rust-lang/rfcs/pull/2603#issuecomment-442837970.